Add docker configuration

arjunsalyan opened the pull request:
- [x] Add docker cofiguration
- [ ] Adjust the app to use new `portindex.json` format
- [ ] Enable updates for the app

Commit:
Add docker configuration

Commit:
Add instructions for running the docker container

umeshksingla commented on README.md:
```diff
@@ -1,4 +1,42 @@
+the docker container. Both way are discussed below:
+
+## 1. Run inside Docker Container
+It is the recommended way, the Docker Image of the app is pre-configured to run the migrations, populate the database and
```
I couldn't locate where it is "pre-configured to run the migrations".

Rest of the PR looks good to me. This document can be improved over time of course.

arjunsalyan replied:
Sorry, I have just written what the docker image would contain. Currently, it does not either run the updates or automatically populate the database. Would this be the wrong way?

But `migrations` do not need to wait. I can add them immediately. Thanks

**UPDATE**: I have added `migration` commands

umeshksingla replied:
I am sorry, I took the readme to describe what has been done; instead of thinking about the doc being work-in-progress on its own.

umeshksingla commented on Dockerfile:
```diff
@@ -0,0 +1,43 @@
+# COPY requirements.txt and RUN pip install BEFORE adding the rest of your code, this will cause Docker's caching mechanism
+# to prevent re-installing (all your) dependencies when you made a change a line or two in your app.
+
+COPY app/requirements.txt /home/docker/code/app/
```
We can really just copy to `/code` or `/src` or `/app` instead of `/home/docker/code/app/`, right?

arjunsalyan replied:
Yes. Does `/home/code` sound better?

Or simply `/home`? Because the Django files would then reside in `/home/app`?

umeshksingla replied:
`/home` may confuse one with the home directory already in ubuntu. You can simply go with `/code`, then the Django files will actually reside in `/code/app`.

arjunsalyan replied:
Okay, thanks. I will change the directory and then we can probably merge this one!

umeshksingla commented:
> Can you make a separate PR for this, so that it can be reviewed and tested separately? You can still have some workspace branch with everything in, and then cherry pick independent items into PRs.

Exactly what I was about to write. We do not want to end up in the same situation again. I think we can even merge this PR already, like now. 

cc @mojca

arjunsalyan commented:
> Can you make a separate PR for this, so that it can be reviewed and tested separately? You can still have some workspace branch with everything in, and then cherry pick independent items into PRs.

I was getting the same feeling, thanks.

umeshksingla commented:
@arjunsalyan Your commits are not getting linked to your Github account. You may want to configure your local Git settings properly. What you can do is add your macports email in Github along with your primary email and configure any of the emails in git config on your machine.

[Link](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-config)

arjunsalyan commented:
> @arjunsalyan Your commits are not getting linked to your Github account. You may want to configure your local Git settings properly. What you can do is add your macports email in Github along with your primary email and configure any of the emails in git config on your machine.
> 
> [Link](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-config)

Thank you, done. Also set my name as "Arjun Salyan" as suggested by @mojca yesterday.

mojca commented on config/uwsgi.ini:
```diff
@@ -0,0 +1,8 @@
+master = true
+processes = 4
+# allow anyone to connect to the socket. This is very permissive
+chmod-socket=666
```
(Nitpicking, but I would try to be consistent with spacing around the equal sign.)

I don't know whether we would need to be stricter with permissions here.

umeshksingla replied:
I am not sure either. Maybe we should go with 664 which seems "more sensible" as mentioned in [uwsgi docs](https://uwsgi-docs.readthedocs.io/en/latest/tutorials/Django_and_nginx.html?highlight=666#if-that-doesn-t-work)?

mojca commented on config/supervisor.conf:
```diff
@@ -0,0 +1,16 @@
+[program:cron]
+autostart=true
+autorestart=true
+command= /bin/bash -c  "declare -p | grep -Ev 'BASHOPTS|BASH_VERSINFO|EUID|PPID|SHELLOPTS|UID' > /code/cron.env && cron -f"
```
Try to use consistent spacing around the equal sign.

arjunsalyan committed:
Make the necessary setup commands separate and run them automatically

Commands like migrate and collectstatic should always run when the
container is started/ restarted. Other commands like load and cron
should be started manually.

arjunsalyan committed:
Allow the app to run from all hosts

arjunsalyan committed:
Make crons to be started manually

umeshksingla commented on Dockerfile:
```diff
@@ -0,0 +1,43 @@
+COPY . /code/
+
+EXPOSE 80
+CMD ["supervisord", "-n"]
```
Probably `/usr/bin/supervisord`?

arjunsalyan replied:
Yes, that can also be used. And would be better, thank you.

umeshksingla commented on config/supervisor.conf:
```diff
@@ -0,0 +1,13 @@
+[program:app-setup]
+command = /bin/bash /code/config/setup_app.sh
+
+[program:cron]
```
Honestly, I have no experience with `supervisord`. I think I got some idea from https://docs.docker.com/config/containers/multi-service_container/ in context of docker. But I have no clue as to what we are trying to achieve with this block here?

arjunsalyan replied:
It creates a file with env variables (which the Cron Jobs can access) and then starts the cron jobs.

umeshksingla replied:
Okay.

umeshksingla replied:
Okay.

umeshksingla commented on config/setup_app.sh:
```diff
@@ -0,0 +1,9 @@
+
+python3 manage.py makemigrations
+
+python3 manage.py migrate
```
The problem with writing a bash script this way is that even if the previous command fails it will throw an error but will simply proceed to execute next command in the line without caring about the result of the last executed command.

arjunsalyan replied:
That is why I provided the most basic commands to this. `migrations` should not cause cause errors that are not known. And in this case it would also not hurt to run any of the three when any failed.

This is just what I feel, please suggest.

umeshksingla replied:
Wrapping the three commands with And operators should suffice.
```
c1 &&
c2 &&
c3
```
Or
```
python3 manage.py makemigrations
[ $? -ne 0 ] || python3 manage.py migrate
[ $? -ne 0 ] || python3 manage.py collectstatic
```

umeshksingla commented on config/crons:
```diff
@@ -0,0 +1,4 @@
+SHELL=/bin/bash
+BASH_ENV=/code/cron.env
+
+0 * * * * root python3 /code/app/manage.py fetch-build-history >> /var/log/cron.log
```
I wouldn't write logs to a file which is a default file for an option made configurable by `cron` for writing CRON entries.

umeshksingla commented on README.md:
````diff
@@ -1,4 +1,52 @@
+
+Now run the image:
+```
+docker run -d -p 80:80 --env-file=env macports-webapp
````
There can be a ton of options we might need to add here later but one option which comes to my mind is `--log-opt max-size=1g` to remove the older logs and stop filling up the disk space on the server.
[Link](https://docs.docker.com/config/containers/logging/json-file/).

arjunsalyan replied:
Yes, thank you.

umeshksingla commented on Dockerfile:
```diff
@@ -0,0 +1,43 @@
+        nginx \
+        cron \
+        supervisor && \
+        pip3 install -U pip setuptools && \
```
It should be moved 4 spaces to the left. But do we even need this?

arjunsalyan replied:
Upgrading `pip` and `setuptools` might be a good idea?

umeshksingla replied:
I was thinking since the script installs pip and setuptools just above this, we probably don't need to upgrade or do anything.

arjunsalyan replied:
I just checked the build logs:
```
Installing collected packages: pip, setuptools
Found existing installation: pip 9.0.1
Not uninstalling pip at /usr/lib/python3/dist-packages, outside environment /usr
Found existing installation: setuptools 39.0.1
Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
Successfully installed pip-19.1.1 setuptools-41.0.1
```
Seems like this command is actually upgrading. Maybe `apt-get` does not install the latest version, so we upgrade using `pip3`?

umeshksingla replied:
I just checked myself and it is indeed upgrading. You can ignore the comment now. Thank you,

umeshksingla commented on Dockerfile:
```diff
@@ -0,0 +1,43 @@
+        cron \
+        supervisor && \
+        pip3 install -U pip setuptools && \
+    rm -rf /var/lib/apt/lists/*
```
Again, any reasons for doing so?

arjunsalyan replied:
I am not very experienced in Linux. But I have seen this in many Dockerfiles.

mojca replied:
Makes sense to me. See https://askubuntu.com/questions/179955/
This is just saving a substantial amount of space on the image.

umeshksingla replied:
Okay. We'll hardly need to do an `apt update` on a running container.

umeshksingla commented on README.md:
````diff
@@ -13,7 +61,7 @@ DATABASES = {
+### Make Migrations
  - Make migrations
  ```
  python manage.py makemigrations
````
`"python3" manage.py ...` maybe? I don't know..

arjunsalyan replied:
Thank you. That would be better.

arjunsalyan committed:
Dockerfile: Minor formatting and use /usr/bin/supervisord

arjunsalyan committed:
Write logs of build history cron to buildhistorycron.log

arjunsalyan committed:
Remove production-only CORS configuration

mojca commented on README.md:
```diff
@@ -1,4 +1,52 @@
+DB_USER=
+DB_PASSWORD=
+DB_HOST=
+SECRET_KEY=
```
We could prepare `env.sample` and ask users to copy that file and fill in the blanks. It's ok to keep the variables documented here as they are now.

We need to document how to run with a local database running on the host machine.

arjunsalyan replied:
Thank you. I will add the file.

Once, I am finish collecting the bits about the entire load and update process, I will prepare a thorough documentation to run the docker image.

arjunsalyan replied:
I have done some changes to the README file. I think it would make sense to add more instructions in the other PR.

umeshksingla replied:
Actually, I'd recommend adding `PORT` to be configured using environment variables as well, thereby completely removing the need to change anything in `settings`, am I right? 
Port is really a part of the host.

arjunsalyan replied:
Okay, thanks. Makes sense to put port in `env`.

mojca commented on README.md:
```diff
@@ -1,4 +1,52 @@
+
+## 2. Run Without Docker
+The `/app` directory is a standalone Django-app which can be run normally like any other django application.
+### Setup the Database
```
Some space before titles?

mojca commented on README.md:
```diff
@@ -1,4 +1,52 @@
+## 2. Run Without Docker
+The `/app` directory is a standalone Django-app which can be run normally like any other django application.
+### Setup the Database
 Enter your database credentials in MacPorts/settings.py
```
This might have a consequence of the user ending up with merge conflicts, or accidentally committing some passwords.

mojca reviewed (commented):
Some minor comments inline. Apart from potential security-related issues (in which I'm no big expert) this mostly look fine. Let me try to test this.

I'm not sure how this is usually done. But the way it's implemented now and with @arjunsalyan apparently uploading the image somewhere ... won't our environmental variables be leaked into that image? What's the best way to define them after creating the base image somehow?

mojca commented on app/MacPorts/settings.py:
```diff
@@ -27,9 +27,7 @@
-    'frozen-falls-98471.herokuapp.com',
-    'localhost',
-    '127.0.0.1',
+    '*'
```
Looking for feedback from @rajdeepbharati.

rajgoesout replied:
Do I need to set up a postgres server to test this?

rajgoesout replied:
I am getting this error:
![Screenshot 2019-07-12 at 2 53 03 PM](https://user-images.githubusercontent.com/30660439/61117745-d1a79480-a4b4-11e9-9932-4cca554edc1f.png)

arjunsalyan replied:
Sorry for late reply. You are probably trying to connect to the database running outside docker container and providing the host `localhost`. A docker container does not have access to the host machine's database through localhost.

I have never tried this, but on Docker for Mac, instead of `localhost` you can use `host.docker.internal` in ` DB_HOST` to access the host's database.

On Docker for Linux, you can use the argument `--network="host"` while running the container.

rajgoesout replied:
Thank you, it works using `host.docker.internal`. Sorry I'm not an expert in Docker

rajgoesout replied:
When I run `docker exec c52e7c755761 python3 /code/app/manage.py load` it gives this error:
`CommandError: "portindex.json" not found. Make sure "portindex.json" is a valid JSON file and is in the root of the project` although I have placed `portindex.json` in `app` directory (also tried keeping it in the root dir)

arjunsalyan replied:
Did you pull the image from docker hub? If yes, you can use the command `load-rsync` instead of `load`.

I am sorry for improper instructions, the setup is being changed and improved regularly, but I will write proper setup instructions soon. For now, using `load-rsync` will solve this.

arjunsalyan replied:
Another better way to do this would be:

```
docker exec <containerid> supervisorctl start app-load
```
This would start the command `load-rsync` and you won't have to worry about keeping the terminal open for 15-20 minutes.

But you should pull the image from docker hub.

```
docker pull arjunsalyan/macports-webapp
```

rajgoesout replied:
Thanks 😄

arjunsalyan replied:
How did you start the command? If you ran

```
docker exec <id> python3 /code/app/manage.py load-rsync
```
then you would have to keep the command running. It should run for 15-20 minutes, there will be some warnings about dependencies, you can ignore them.

But if you use
```
docker exec <id> supervisorctl start app-load
```

then you won't have to worry about anything.

rajgoesout replied:
I used this: `docker exec <id> python3 /code/app/manage.py load-rsync`

arjunsalyan replied:
Okay, the command should have finished by now.

arjunsalyan commented:
> Some minor comments inline. Apart from potential security-related issues (in which I'm no big expert) this mostly look fine. Let me try to test this.
> 
> I'm not sure how this is usually done. But the way it's implemented now and with @arjunsalyan apparently uploading the image somewhere ... won't our environmental variables be leaked into that image? What's the best way to define them after creating the base image somehow?

All the environment variables are in the file `env` that is supplied to the image while running the container. The image on docker hub is built using this repository. The image does not know anything about environment variables until the file env is supplied to it. I do not think there is any security issue with this.

The uploaded image has nothing more than what is on this repository. The image is then pulled and supplied with variables when running.

Did I get your point right? Sorry, if you were trying to say something else.

arjunsalyan committed:
Format the spacing around = in config files

arjunsalyan committed:
Add file env.sample

arjunsalyan committed:
Update and format README.md

arjunsalyan committed:
Add --noinput argument to initial commands

arjunsalyan committed:
Dockerfile: Add "rsync" to the install command

umeshksingla commented:
> > Some minor comments inline. Apart from potential security-related issues (in which I'm no big expert) this mostly look fine. Let me try to test this.
> > I'm not sure how this is usually done. But the way it's implemented now and with @arjunsalyan apparently uploading the image somewhere ... won't our environmental variables be leaked into that image? What's the best way to define them after creating the base image somehow?
> 
> All the environment variables are in the file `env` that is supplied to the image while running the container. The image on docker hub is built using this repository. The image does not know anything about environment variables until the file env is supplied to it. I do not think there is any security issue with this.
> 
> The uploaded image has nothing more than what is on this repository. The image is then pulled and supplied with variables when running.
> 
> Did I get your point right? Sorry, if you were trying to say something else.

Unless you spotted the password, host, etc explicitly mentioned somewhere in the code, the image doesn't contain any reference to the values and whenever the container is run from the image, the user would need to specify them, ie `docker run -e A=1 -e B=2 <image_name>` using CLI or env file as in readme. The container will crash if any of these variables is not specified.
That being said, I'm a little behind on reviewing how Travis is handling the db operations.

arjunsalyan commented:
> That being said, I'm a little behind on reviewing how Travis is handling the db operations.

For Travis, these variables are defined in `.travis.yml`. https://github.com/arjunsalyan/macports-gsoc-2019-webapp/blob/452c080be083975cbc7d91c2a30013a356b0bf2d/.travis.yml#L13

umeshksingla commented:
I tried running the app without specifying the env credentials and this is what I get:
```
umeshksingla@Umeshs-MacBook-Pro-2:macports-gsoc-2019-webapp$ docker run mp-wa
/usr/lib/python2.7/dist-packages/supervisor/options.py:298: UserWarning: Supervisord is running as root and it is searching for its configuration file in default locations (including its current working directory); you probably want to specify a "-c" argument specifying an absolute path to a configuration file for improved security.
  'Supervisord is running as root and it is searching '
2019-07-14 12:22:11,151 CRIT Supervisor running as root (no user in config file)
2019-07-14 12:22:11,151 INFO Included extra file "/etc/supervisor/conf.d/supervisor.conf" during parsing
2019-07-14 12:22:11,159 INFO RPC interface 'supervisor' initialized
2019-07-14 12:22:11,159 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2019-07-14 12:22:11,159 INFO supervisord started with pid 1
2019-07-14 12:22:12,163 INFO spawned: 'app-uwsgi' with pid 9
2019-07-14 12:22:12,166 INFO spawned: 'nginx-app' with pid 10
2019-07-14 12:22:12,168 INFO spawned: 'app-setup' with pid 11
2019-07-14 12:22:12,385 INFO exited: app-setup (exit status 1; not expected)
2019-07-14 12:22:13,421 INFO success: app-uwsgi entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-07-14 12:22:13,422 INFO success: nginx-app entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2019-07-14 12:22:13,423 INFO spawned: 'app-setup' with pid 24
2019-07-14 12:22:13,626 INFO exited: app-setup (exit status 1; not expected)
2019-07-14 12:22:15,633 INFO spawned: 'app-setup' with pid 28
2019-07-14 12:22:15,833 INFO exited: app-setup (exit status 1; not expected)
2019-07-14 12:22:18,845 INFO spawned: 'app-setup' with pid 32
2019-07-14 12:22:19,055 INFO exited: app-setup (exit status 1; not expected)
2019-07-14 12:22:20,058 INFO gave up: app-setup entered FATAL state, too many start retries too quickly
```
and eventually exits the container which is right. I tried checking the actual service logs using `docker logs <container-id>` but it had the same output as above. There is really no way to see what went wrong until I started the container with `docker run -it <image> bash` and ran `/bin/bash /code/config/setup_app.sh` myself and it finally printed out the error I have been looking for:
```
Traceback (most recent call last):
  File "manage.py", line 15, in <module>
....
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/code/app/MacPorts/settings.py", line 24, in <module>
    SECRET_KEY = os.environ['SECRET_KEY']
  File "/usr/lib/python3.6/os.py", line 669, in __getitem__
    raise KeyError(key) from None
KeyError: 'SECRET_KEY'
```
[This link](http://veithen.io/2015/01/08/supervisord-redirecting-stdout.html) helps with capturing the output of docker container to supervisor's stdout. So this config worked for me and I got the error logs from container on simply running the image:
```
[program:app-setup]
command = /bin/bash /code/config/setup_app.sh
stdout_logfile=/dev/stdout
stdout_logfile_maxbytes=0
stderr_logfile=/dev/stderr
stderr_logfile_maxbytes=0
```

arjunsalyan committed:
Pass DB_PORT as environment variable

arjunsalyan commented:
> [program:app-setup]
> command = /bin/bash /code/config/setup_app.sh
> stdout_logfile=/dev/stdout
> stdout_logfile_maxbytes=0
> stderr_logfile=/dev/stderr
> stderr_logfile_maxbytes=0

I think we should make this default. It would be useful to have the output of these setup commands printed.

mojca commented:
> Unless you spotted the password, host, etc explicitly mentioned somewhere in the code, the image doesn't contain any reference to the values and whenever the container is run from the image, the user would need to specify them, ie `docker run -e A=1 -e B=2 <image_name>` using CLI or env file as in readme. The container will crash if any of these variables is not specified.

When I first tested the docker setup (weeks ago) I remember seeing one command that copied env somewhere, and that one of the steps failed if the env file was missing. I'm no longer sure if that was in fact a problem, or I simply had troubles running (not creating) the image itself due to the lack of the `env` file. So ignore that comment of time, it seems to be ok.

mojca commented on README.md:
```diff
@@ -1,45 +1,106 @@
+DB_NAME=
+DB_USER=
+DB_PASSWORD=
+DB_HOST=
```
Just below these variables we could add a short sentence saying something like this:

> If you want to run the database on the host running macOS, you may set `DB_HOST=host.docker.internal`, [(link)](https://nickjanetakis.com/blog/docker-tip-65-get-your-docker-hosts-ip-address-from-in-a-container)

arjunsalyan replied:
I have added this in the other PR. But it would be useful to have it here.

mojca reviewed (approved):
I didn't start the cron jobs etc., but until that step the instructions now work perfect, and I could easily connect to the database on host using `DB_HOST=host.docker.internal`, and the application runs, accepts statistics etc.

My suggestion would be to squash the changes and merge this.

umeshksingla commented:
It worked for me too on localhost. I tried running fetch-build-history on the empty database of ports and it gave me this:
```
root@bd9df13a3674:/code/app# python3 manage.py fetch-build-history
/usr/local/lib/python3.6/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
root@bd9df13a3674:/code/app# ls
```

It's getting late for me at the moment for office so I'll try to debug a little later in the day. I have a doubt around how the list of builders gets loaded in the db because the `builders` table is empty right after migrations and the first line of `BuildHistory.populate()` is:
`builders = Builder.objects.values_list('name', flat=True)`. 

What am I doing wrong?

arjunsalyan commented:
> It worked for me too on localhost. I tried running fetch-build-history on the empty database of ports and it gave me this:
> 
> ```
> root@bd9df13a3674:/code/app# python3 manage.py fetch-build-history
> /usr/local/lib/python3.6/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
>   """)
> root@bd9df13a3674:/code/app# ls
> ```
> 
> It's getting late for me at the moment for office so I'll try to debug a little later in the day. I have a doubt around how the list of builders gets loaded in the db because the `builders` table is empty right after migrations and the first line of `BuildHistory.populate()` is:
> `builders = Builder.objects.values_list('name', flat=True)`.
> 
> What am I doing wrong?

This is happening because the builders table is empty. The command fetches history only for those builders which are present in the table. We could automate the process of adding the builders to the table. But I am not sure of the right place to hardcode builders.

Data migrations could be one way, but we probably do not want to automate the process of populating the database. Then the right way seems to be adding a `bulk_create` inside the command `autoload-initial-data`.

arjunsalyan committed:
Add 0001_initial.py migration file

arjunsalyan commented:
@mojca @umeshksingla : Did you pull the image from Docker Hub or built one using the contents of this PR?

The image on Docker Hub is much ahead of this PR- it is being built using the branch [production/docker](https://github.com/arjunsalyan/macports-gsoc-2019-webapp/tree/production/docker).

mojca commented:
I tested building from source to make sure that instructions work. If the rest of functionality has been improved elsewhere, that's fine & fully expected. The important part of testing build from sources is to make sure that your work is reproducible elsewhere. The database and migrations were applied correctly, but I didn't set any of the cron jobs.

umeshksingla commented:
> @mojca @umeshksingla : Did you pull the image from Docker Hub or built one using the contents of this PR?
> 
> The image on Docker Hub is much ahead of this PR- it is being built using the branch [production/docker](https://github.com/arjunsalyan/macports-gsoc-2019-webapp/tree/production/docker).

I built from source since I wanted to check if it the commands run fine and also check the supervisor instructions. I wasn't able to get the cron running as above. I'll try tonight again.

arjunsalyan commented:
> I tested building from source to make sure that instructions work. If the rest of functionality has been improved elsewhere, that's fine & fully expected. The important part of testing build from sources is to make sure that your work is reproducible elsewhere. The database and migrations were applied correctly, but I didn't set any of the cron jobs.

Thanks for testing. The cron would start but build history won't be fetched as there are no builders in the table yet.

arjunsalyan commented:
> > @mojca @umeshksingla : Did you pull the image from Docker Hub or built one using the contents of this PR?
> > The image on Docker Hub is much ahead of this PR- it is being built using the branch [production/docker](https://github.com/arjunsalyan/macports-gsoc-2019-webapp/tree/production/docker).
> 
> I built from source since I wanted to check if it the commands run fine and also check the supervisor instructions. I wasn't able to get the cron running as above. I'll try tonight again.

Thank you. I need to fix adding builders to the table by then.
